### PR TITLE
[Bugfix][Kernel] Fix persistent top-k histogram reuse after short rows

### PR DESCRIPTION
**Pull request by fxfxfxfxfxfxfxfx**
## Purpose

Fix a correctness bug in `persistent_topk` when one persistent CTA group
processes a radix row (`seq_len > 32768`), followed by a short or medium row
(`seq_len <= 32768`), and then another radix row.

The kernel previously used the outer row iteration counter to rotate its
triple-buffered radix histograms. Short rows advanced that counter without
executing `radix_topk`, causing the next radix row to reuse a histogram that
could still contain counts from an earlier row.

This PR introduces a separate `radix_iter` counter that advances only when
`radix_topk` executes. It also adds a CUDA regression test that constructs the
long-short-long scheduling pattern in one persistent CTA group.

Related to #41748, which addresses workspace initialization between kernel
launches. This PR fixes histogram-ring misalignment within a single launch.

## Test Plan

- Compare the long-short-long case against `torch.topk`.
- Run a consecutive-radix case as a control.
- Repeat both cases ten times on an RTX 3080.
- Exercise the adaptive CTA-group layout used by the regression test.
- Run the checked-in regression test in CI:

```bash
python -m pytest \
  tests/kernels/test_top_k_per_row.py::test_persistent_topk_reused_group_after_short_row \
  -v
```

- Run static checks:

```bash
python -m py_compile tests/kernels/test_top_k_per_row.py
ruff check tests/kernels/test_top_k_per_row.py
git diff --check origin/main...HEAD
git clang-format --diff origin/main -- \
  csrc/libtorch_stable/persistent_topk.cuh
```

## Test Result

RTX 3080 results before the fix:

```text
middle_len=32768: match=False, set_diff=8
middle_len=32769: match=True,  set_diff=0
```

Results after rebuilding the production `topk.cu` with the fix:

```text
middle_len=32768: failures=0/10
middle_len=32769: failures=0/10
```

Adaptive regression layout:

```text
ctas_per_group=7, num_groups=9, rows=(0, 9, 18)
set_diff=0
```

Python compilation, Ruff, `git diff --check`, and clang-format checks passed.

The complete `_C_stable_libtorch` extension and checked-in pytest were not run
locally; they are left for CI. Local GPU validation compiled the production
`topk.cu` and modified `persistent_topk.cuh` for SM86.

No documentation update is required because this is an internal kernel
correctness fix with no public API or configuration changes.

AI assistance disclosure: OpenAI Codex was used for source analysis,
implementation, regression-test construction, and validation. I reviewed and
understand all submitted changes.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results.
- [x] No documentation update is required for this internal kernel fix.

</details>

**BEFORE SUBMITTING, PLEASE READ <https://docs.vllm.ai/en/latest/contributing>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by fxfxfxfxfxfxfxfx** [comment]
Gentle ping for review @dcampora @LopezCastroRoberto — this fixes a correctness bug in `persistent_topk` that could silently return wrong top-k results.

**Context**: when one persistent CTA group schedules a radix row (seq_len > 32768), then a short/medium row, then another radix row, the outer row counter rotates the triple-buffered histograms even though `radix_topk` never ran for the short row. The next radix row then reuses a histogram that still holds counts from an earlier row — no crash, just wrong counts, which makes it easy to miss.

**Fix**: a separate `radix_iter` counter that only advances when `radix_topk` actually executes, plus a CUDA regression test that builds the long-short-long pattern in one persistent CTA group (verified 10/10 pass on RTX 3080; before the fix it produced `set_diff=8`).

This complements #41748 (workspace init between launches) — this PR fixes histogram-ring misalignment *within* a single launch. @dcampora since you wrote the original kernel, and @LopezCastroRoberto since it interacts with your workspace work, your eyes on the radix logic would be especially valuable. Happy to split or rebase if that helps review.

**Comment by fxfxfxfxfxfxfxfx** [comment]
@mgoin

**Review by LopezCastroRoberto — APPROVED** [review]
LGTM, thanks for the fix!

**Comment by LopezCastroRoberto** [comment]
@fxfxfxfxfxfxfxfx can you please fix the pre-commit?

**Review by mgoin — APPROVED** [review]

**Comment by fxfxfxfxfxfxfxfx** [comment]
@mgoin I checked the full log, and this does not appear to be related to this PR. Both the original run and the retry crash during model warmup with `CUDBG_EXCEPTION_WARP_OUT_OF_RANGE_ADDRESS` in `deep_gemm::sched::sm90_paged_mqa_logits_metadata<1024, 256, 132, false>`, while this PR only changes `persistent_topk`. Could you please take a look?

**Comment by zyongye** [comment]
> @mgoin I checked the full log, and this does not appear to be related to this PR. Both the original run and the retry crash during model warmup with `CUDBG_EXCEPTION_WARP_OUT_OF_RANGE_ADDRESS` in `deep_gemm::sched::sm90_paged_mqa_logits_metadata<1024, 256, 132, false>`, while this PR only changes `persistent_topk`. Could you please take a look?

yea it is not related. We are fixing it rn.
